File: files/nvim/lua/lusty/fuzzy.lua
-- Smart subsequence fuzzy scorer (fzy-style) for LustyExplorer.
--
-- Used by default for ranking matches while typing; the original Mercury
-- scorer stays available via g:LustyExplorerFuzzyEngine = 'mercury'.
--
-- Scoring ideas:
--   * matched characters earn a base score plus a word-boundary bonus when
--     they start a segment (string start, after '/' '-' '_' '.' ' '), or on
--     a camelCase boundary;
--   * consecutive matched characters in the haystack get a small bonus;
--   * skipped characters between matches cost a little (inner gap), and a
--     late first match is penalised, so prefix/segment-leading matches win.

local M = {}

local LEADING_PENALTY = 0.012 -- per skipped char before the first match
local INNER_GAP = 0.02       -- per skipped char between two matches
local CONSECUTIVE_BONUS = 0.05

local function is_upper(b)
  return b >= 65 and b <= 90
end
local function is_lower(b)
  return b >= 97 and b <= 122
end

--- Boundary bonus for a haystack byte position (1-based).
local function bonus_for(str, i)
  if i == 1 then
    return 0.9
  end
  local ch = str:sub(i, i)
  if ch == '/' then
    return 0.9
  end
  local prev = str:sub(i - 1, i - 1)
  if prev == '-' or prev == '_' or prev == ' ' or prev == '.' then
    return 0.8
  end
  if prev == '/' then
    return 0.9
  end
  -- camelCase: lower-case letter followed by an upper-case one.
  local pb = str:byte(i - 1, i - 1)
  local cb = str:byte(i, i)
  if pb and cb and is_lower(pb) and is_upper(cb) then
    return 0.7
  end
  return 0.0
end

-- Subsequence check (case-insensitive).
local function has_match(lower_str, lower_abbrev)
  local a, b = 1, 1
  while a <= #lower_abbrev and b <= #lower_str do
    if lower_abbrev:sub(a, a) == lower_str:sub(b, b) then
      a = a + 1
    end
    b = b + 1
  end
  return a > #lower_abbrev
end

--- Score how well `abbrev` matches `str` (case-insensitive).
--- @param str string
--- @param abbrev string
--- @return number (0.0 when there is no match)
function M.score(str, abbrev)
  if abbrev == '' then
    return 0.75 -- neutral score for an empty query
  end
  if #abbrev > #str then
    return 0.0
  end

  local lower_str = str:lower()
  local lower_abbrev = abbrev:lower()
  if not has_match(lower_str, lower_abbrev) then
    return 0.0
  end

  local n = #lower_abbrev
  local m = #lower_str

  -- Rolling DP rows: prev[j] holds the best score for the previous pattern
  -- prefix ending exactly at haystack position j.
  local prev = {}
  local cur = {}
  local best = -math.huge

  for i = 1, n do
    -- Running max of (prev[k] + INNER_GAP * k) over k < j enables the
    -- inner-gap transition in O(1) per cell.
    local best_gap = -math.huge
    for j = 1, m do
      local score_j = nil
      if lower_abbrev:sub(i, i) == lower_str:sub(j, j) then
        local bon = bonus_for(str, j)
        if i == 1 then
          score_j = bon - LEADING_PENALTY * (j - 1)
        else
          local cand = -math.huge
          if prev[j - 1] ~= nil then
            cand = prev[j - 1] + CONSECUTIVE_BONUS
          end
          if best_gap > -math.huge then
            local via_gap = best_gap - INNER_GAP * (j - 1)
            if via_gap > cand then
              cand = via_gap
            end
          end
          if cand > -math.huge then
            score_j = cand + bon
          end
        end
      end
      cur[j] = score_j
      if score_j ~= nil and i == n and score_j > best then
        best = score_j
      end
      -- Allow prev[j] to serve gap transitions for positions after j.
      if prev[j] ~= nil then
        local with_k = prev[j] + INNER_GAP * j
        if with_k > best_gap then
          best_gap = with_k
        end
      end
    end
    prev, cur = cur, prev
  end

  if best <= -math.huge / 2 then
    return 0.0
  end
  return best
end

return M

File: packages/lusty-native/src/fuzzy.rs
//! Smart subsequence fuzzy scorer (fzy-style), ported 1:1 from the Lua port
//! (files/nvim/lua/lusty/fuzzy.lua). Default engine for Lusty-native.

const LEADING_PENALTY: f64 = 0.012; // per skipped char before the first match
const INNER_GAP: f64 = 0.02; // per skipped char between two matches
const CONSECUTIVE_BONUS: f64 = 0.05;

/// Boundary bonus for a haystack byte position (0-based).
fn bonus_for(s: &[u8], i: usize) -> f64 {
    if i == 0 {
        return 0.9;
    }
    let ch = s[i];
    if ch == b'/' {
        return 0.9;
    }
    let prev = s[i - 1];
    if prev == b'-' || prev == b'_' || prev == b' ' || prev == b'.' {
        return 0.8;
    }
    if prev == b'/' {
        return 0.9;
    }
    // camelCase: lower-case letter followed by an upper-case one.
    if prev.is_ascii_lowercase() && ch.is_ascii_uppercase() {
        return 0.7;
    }
    0.0
}

/// Case-insensitive subsequence check (both slices already lowercased).
fn has_match(lower_s: &[u8], lower_a: &[u8]) -> bool {
    let (mut a, mut b) = (0usize, 0usize);
    while a < lower_a.len() && b < lower_s.len() {
        if lower_a[a] == lower_s[b] {
            a += 1;
        }
        b += 1;
    }
    a == lower_a.len()
}

/// Reusable scorer. Ranking thousands of entries on every keystroke must not
/// allocate two DP rows plus a lowercase copy per candidate, so the scratch
/// buffers live here and are only grown when a longer label shows up.
pub struct Scorer {
    prev: Vec<f64>,
    cur: Vec<f64>,
    hay: Vec<u8>,
}

impl Default for Scorer {
    fn default() -> Self {
        Self::new()
    }
}

impl Scorer {
    pub fn new() -> Scorer {
        Scorer {
            prev: Vec::new(),
            cur: Vec::new(),
            hay: Vec::new(),
        }
    }

    /// Score how well `abbrev` matches `str_` (case-insensitive); 0.0 means
    /// no subsequence match.
    pub fn score(&mut self, str_: &str, abbrev: &str) -> f64 {
        if abbrev.is_empty() {
            return 0.75; // neutral score for an empty query
        }
        let s = str_.as_bytes();
        let a = abbrev.as_bytes();
        if a.len() > s.len() {
            return 0.0;
        }
        self.hay.clear();
        self.hay.extend(s.iter().map(|b| b.to_ascii_lowercase()));
        let m = self.hay.len();
        let n = a.len();
        // Reuse the DP rows; grow only when this label is longer than before.
        for row in [&mut self.prev, &mut self.cur] {
            if row.len() < m + 1 {
                row.resize(m + 1, f64::NEG_INFINITY);
            }
            row[..m + 1].fill(f64::NEG_INFINITY);
        }
        let lower_a: Vec<u8> = a.iter().map(|b| b.to_ascii_lowercase()).collect();
        if !has_match(&self.hay, &lower_a) {
            return 0.0;
        }

        let neg = f64::NEG_INFINITY;
        let (prev, cur) = (&mut self.prev, &mut self.cur);
        let mut best = neg;

        for i in 1..=n {
            // Running max of (prev[k] + INNER_GAP * k) over k < j enables the
            // inner-gap transition in O(1) per cell.
            let mut best_gap = neg;
            for j in 1..=m {
                let mut score_j = neg;
                if lower_a[i - 1] == self.hay[j - 1] {
                    let bon = bonus_for(s, j - 1);
                    if i == 1 {
                        score_j = bon - LEADING_PENALTY * (j - 1) as f64;
                    } else {
                        let mut cand = neg;
                        if prev[j - 1].is_finite() {
                            cand = prev[j - 1] + CONSECUTIVE_BONUS;
                        }
                        if best_gap.is_finite() {
                            let via_gap = best_gap - INNER_GAP * (j - 1) as f64;
                            if via_gap > cand {
                                cand = via_gap;
                            }
                        }
                        if cand.is_finite() {
                            score_j = cand + bon;
                        }
                    }
                }
                cur[j] = score_j;
                if score_j.is_finite() && i == n && score_j > best {
                    best = score_j;
                }
                // Allow prev[j] to serve gap transitions for positions after j.
                if prev[j].is_finite() {
                    let with_k = prev[j] + INNER_GAP * j as f64;
                    if with_k > best_gap {
                        best_gap = with_k;
                    }
                }
            }
            std::mem::swap(prev, cur);
        }

        if !best.is_finite() {
            return 0.0;
        }
        best
    }
}

/// One-shot convenience wrapper (allocates scratch per call); used by the
/// reference tests and handy for callers that score a single pair.
#[allow(dead_code)]
pub fn score(str_: &str, abbrev: &str) -> f64 {
    Scorer::new().score(str_, abbrev)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_query_neutral() {
        assert_eq!(score("anything", ""), 0.75);
    }

    #[test]
    fn boundary_beats_midword_leading_match() {
        // 'b' at the string start earns the start bonus; a mid-word 'c' with
        // no boundary scores negative (the Lua port returns -0.024 here).
        let prefix = score("beta.lua", "b");
        let inner = score("pic.jpg", "c");
        assert_eq!(prefix, 0.9);
        assert!(prefix > inner);
        assert!(inner < 0.0);
    }

    #[test]
    fn word_boundary_bonus() {
        // matching 'g' after the slash in sub/gamma.txt is a strong boundary
        let after_slash = score("sub/gamma.txt", "g");
        let mid_word = score("sugar.txt", "g");
        assert!(after_slash > mid_word);
        assert!(after_slash > 0.0);
        assert!(mid_word < 0.0);
    }

    #[test]
    fn boundary_bonus_can_beat_consecutive() {
        // 'at' lands on the '.'-boundary ('t' after '.'), so it outscores the
        // plain consecutive prefix 'alp' -- matches the Lua port exactly.
        let consecutive = score("alpha.txt", "alp");
        let gapped = score("alpha.txt", "at");
        assert_eq!(consecutive, 1.0);
        assert_eq!(gapped, 1.6);
        assert!(gapped > consecutive);
    }

    #[test]
    fn case_insensitive() {
        assert_eq!(score("Alpha.txt", "alp"), score("alpha.txt", "alp"));
        assert!(score("Alpha.txt", "ALP") > 0.0);
    }

    #[test]
    fn no_match_returns_zero() {
        assert_eq!(score("alpha.txt", "zzz"), 0.0);
        assert_eq!(score("beta.lua", "alpha"), 0.0);
    }

    #[test]
    fn longer_query_than_string() {
        assert_eq!(score("ab", "abc"), 0.0);
    }

    #[test]
    fn camel_case_boundary() {
        let cc = score("MyUgens.sc", "mu");
        let plain = score("myugens.sc", "mu");
        assert!(cc > plain);
    }

    #[test]
    fn port_matches_lua_reference_table() {
        // Ground-truth values printed by the real Lua port
        // (nvim --clean --headless -l, files/nvim/lua/lusty/fuzzy.lua).
        let cases: &[(&str, &str, f64)] = &[
            ("pic.jpg", "c", -0.024),
            ("beta.lua", "b", 0.9),
            ("beta.lua", "beta", 1.05),
            ("alpha.txt", "alp", 1.0),
            ("alpha.txt", "at", 1.6),
            ("sub/gamma.txt", "g", 0.852),
            ("sugar.txt", "g", -0.024),
            ("sub/gamma.txt", "gamma", 1.052),
            ("mug.png", "gamma", 0.0),
            ("Alpha.txt", "alp", 1.0),
            ("Alpha.txt", "ALP", 1.0),
            ("MyUgens.sc", "mu", 1.58),
            ("myugens.sc", "mu", 0.88),
            ("alpha.txt", "zzz", 0.0),
            ("ab", "abc", 0.0),
        ];
        for (s, a, want) in cases {
            let got = score(s, a);
            assert!(
                (got - want).abs() < 1e-9,
                "score({:?}, {:?}) = {} (want {})",
                s,
                a,
                got,
                want
            );
        }
    }
}
